src/object-store/src/layers/lru_cache/read_cache.rs

```diff
@@ -150,24 +150,32 @@ impl<C: Access> ReadCache<C> {
     /// Return entry count and total approximate entry size in bytes.
     pub(crate) async fn recover_cache(&self) -> Result<(u64, u64)> {
         let op = OperatorBuilder::new(self.file_cache.clone()).finish();
+        let cloned_op = op.clone();
         let root = read_cache_root();
         let mut entries = op
-            .list_with(&root)
-            .metakey(Metakey::ContentLength | Metakey::ContentType)
-            .concurrent(RECOVER_CACHE_LIST_CONCURRENT)
+            .lister_with(&root)
+            .await?
+            .map_ok(|entry| async {
+                let (path, mut meta) = entry.into_parts();
+
+                if !cloned_op.info().full_capability().list_has_content_length {
+                    meta = cloned_op.stat(&path).await?;
+                }
+
+                Ok((path, meta))
+            })
+            .try_buffer_unordered(RECOVER_CACHE_LIST_CONCURRENT)
+            .try_collect::<Vec<_>>()
             .await?;
 
-        while let Some(entry) = entries.pop() {
-            let read_key = entry.path();
-            let size = entry.metadata().content_length();
+        while let Some((read_key, metadata)) = entries.pop() {
+            let size = metadata.content_length();
             OBJECT_STORE_LRU_CACHE_ENTRIES.inc();
             OBJECT_STORE_LRU_CACHE_BYTES.add(size as i64);
-            // ignore root path
-            if entry.metadata().mode() == EntryMode::FILE {
-                self.mem_cache
-                    .insert(read_key.to_string(), ReadResult::Success(size as u32))
-                    .await;
-            }
+
+            self.mem_cache
```

Comment:
Do we still need to ignore the root path?

Reply:
Addressed.